### PR TITLE
[Table Visualization] make table vis column sortable

### PR DESCRIPTION
**Pull request by ananzh**
### Description
* add sort state (asc | desc)  to column

### Partially resolve:
https://github.com/opensearch-project/OpenSearch-Dashboards/issues/2305

### Check List
- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/vis_type_table/public/components/table_vis_component.tsx [hunk 1]
```diff
@@ -20,10 +21,38 @@ interface TableVisComponentProps {
...
+    handlers.uiState.get('vis.columnsWidth') || []
+  );
+
+  // store current state
+  const curColumnsState = useRef<{
+    sortColumn: SortColumn;
+    columnsWidth: ColumnWidth[];
+  }>({
+    sortColumn: handlers.uiState.get('vis.sortColumn'),
+    columnsWidth: handlers.uiState?.get('vis.columnsWidth'),
```
Comment (joshuarrrr) [inline src/plugins/vis_type_table/public/components/table_vis_component.tsx:46]: nit - if you actually need `handlers.uiState?.` here, you may need it elsewhere - I'd expect it to be consistent.

File: src/plugins/vis_type_table/public/components/table_vis_component.tsx [hunk 2]
```diff
@@ -20,10 +21,38 @@ interface TableVisComponentProps {
...
   const pagination = usePagination(visConfig, rows.length);
 
+  // set sort and resize column state
+  const [sortColumn, setSortColumn] = useState<SortColumn>(
+    handlers.uiState.get('vis.sortColumn') || {}
+  );
+
+  const [columnsWidth, setColumnsWidth] = useState<ColumnWidth[]>(
+    handlers.uiState.get('vis.columnsWidth') || []
+  );
```
Comment (joshuarrrr) [inline src/plugins/vis_type_table/public/components/table_vis_component.tsx:38]: Do you actually need these local states? It seems like you can just use `handlers.uiState` getters/setters and the `curColumnsState`.

File: src/plugins/vis_type_table/public/components/table_vis_component.tsx [hunk 3]
```diff
@@ -34,49 +63,70 @@ export const TableVisComponent = ({ table, visConfig, handlers }: TableVisCompon
...
-  const [columnsWidth, setColumnsWidth] = useState<ColumnWidth[]>(
-    handlers.uiState.get('vis.columnsWidth') || []
+  }, [sortedRows, pagination]);
+
+  const dataGridColumns = getDataGridColumns(sortedRows, columns, table, handlers, columnsWidth);
+
+  const sortedColumns = useMemo(() => {
+    const sort = handlers.uiState?.get('vis.sortColumn');
+    return sort && sort.colIndex !== null && sort.direction
+      ? [{ id: dataGridColumns[sort.colIndex]?.id, direction: sort.direction }]
```
Comment (joshuarrrr) [inline src/plugins/vis_type_table/public/components/table_vis_component.tsx:77]: If we eventually want `sortedColumns` by `id`, why  does the `SortColumn` type use the column index? I see lots of mappings back and forth between index and id, but it's unclear if the index value matters all that much - maybe just for the column width/resizing?

**Review by joshuarrrr — COMMENTED** [review]
Nice progress. I don't have all the context of how this should be working, but it seems like the state handling could be simplified a bit more.

File: src/plugins/vis_type_table/public/components/table_vis_component.tsx [hunk 4]
```diff
@@ -20,10 +21,38 @@ interface TableVisComponentProps {
...
+  const { formattedRows: rows, formattedColumns: columns } = convertToFormattedData(
+    table,
+    visConfig
+  );
 
   const pagination = usePagination(visConfig, rows.length);
 
+  // set sort and resize column state
+  const [sortColumn, setSortColumn] = useState<SortColumn>(
+    handlers.uiState.get('vis.sortColumn') || {}
```
Comment (ashwin-pc) [inline src/plugins/vis_type_table/public/components/table_vis_component.tsx:33]: From what i know about UIState, its a way to get and set state information inside a component rendered by an expression function. We should not be maintaining duplicate state here. If we want to set the new `sortColumn` value, just use the `uiState.set` function. The Visualization should rerender with the updated value automatically.

File: src/plugins/vis_type_table/public/components/table_vis_component.tsx (hunk 1 above)
Comment (ananzh) [reply src/plugins/vis_type_table/public/components/table_vis_component.tsx:46]: will fix this.

File: src/plugins/vis_type_table/public/components/table_vis_component.tsx (hunk 2 above)
Comment (ananzh) [reply src/plugins/vis_type_table/public/components/table_vis_component.tsx:38]: got it. I will combine these local states

File: src/plugins/vis_type_table/public/components/table_vis_component.tsx (hunk 4 above)
Comment (ananzh) [reply src/plugins/vis_type_table/public/components/table_vis_component.tsx:33]: > 

Hey Ashwin, regarding this feedback (https://github.com/opensearch-project/OpenSearch-Dashboards/pull/2502#discussion_r988409245 ), you are right, uiState.set won’t let me to render the component. If I switch setSortColumn(updatedVisState.sortColumn); to uiState.set than it won’t re-render with corrected sorted columns. I am using an example from charts ([https://github.com/opensearch-project/OpenSearch-Dashboards/blob/d7004dc5b0392477f[…]3b/src/plugins/charts/public/static/components/color_schema.tsx](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/d7004dc5b0392477fdd54ac66b29d231975a173b/src/plugins/charts/public/static/components/color_schema.tsx#L64) ) to force it work. Regarding this part, I could open an issue as a post research topic. Currently, I couldn’t find anything meaningful now.

File: src/plugins/vis_type_table/public/components/table_vis_component.tsx (hunk 3 above)
Comment (ananzh) [reply src/plugins/vis_type_table/public/components/table_vis_component.tsx:77]: Yeah I am aware of this. The reason why I am using index not id is that when we add/remove a column, the id might change. Below is an example that I add another column day_of_week. Count id is changed.

Screen Shot 2022-10-10 at 19 32 13

Screen Shot 2022-10-10 at 19 32 59

I think there might be ways to work around. but for now, using index is the safest way to guarantee we could find the correct col. 

I opened an issue to research on this to see if we could simply the code:
https://github.com/opensearch-project/OpenSearch-Dashboards/issues/2549

I will work on it next sprint.

File: src/plugins/vis_type_table/public/components/table_vis_component.tsx (hunk 4 above)
Comment (ananzh) [reply src/plugins/vis_type_table/public/components/table_vis_component.tsx]: removed local state.

**Pull request merged**